language/move-vm/runtime/src/data_cache.rs

```diff
@@ -44,21 +58,30 @@ impl ConfigStorage for &dyn RemoteCache {
 /// The Move VM takes a `DataStore` in input and this is the default and correct implementation
 /// for a data store related to a transaction. Clients should create an instance of this type
 /// and pass it to the Move VM.
-pub struct TransactionDataCache<'txn> {
-    data_map: BTreeMap<AccessPath, Option<(FatStructType, GlobalValue)>>,
-    module_map: BTreeMap<ModuleId, Vec<u8>>,
-    event_data: Vec<ContractEvent>,
-    data_cache: &'txn dyn RemoteCache,
+pub struct TransactionDataCache<'r, 'l, C> {
+    data_cache: &'r C,
+    loader: &'l Loader,
+    account_map: BTreeMap<AccountAddress, AccountDataCache>,
+    event_data: Vec<(Vec<u8>, u64, Type, Value)>,
```

Comment:
I understand the intent here but I am not really sure that removing `ContractEvent` makes things better

Reply:
`ContractEvent` is now a Libra concept. What I did was essentially mirroring the concept in Move. It might make sense just to promote `ContractEvent` into Move instead. Let's think about whether that's reasonable.

Reply:
Another alternative I discussed with @runtian-zhou was to define 
```
struct LibraContractEvent<T> {
    guid: vector<u8>,
    seq_num: u64,
    msg: T,
}
```
in Move so we can have a much simpler event model in the Move VM. However this does require us to take the Value apart in the Libra VM and construct a ContractEvent there.

Reply:
I understand the intent. 
However my point is that there is no concept of event today in the VM, it's not clear what that would be, it relates to native functions and more.
So conceptually this sounds good but the truth is that we are just deconstructing a `ContractEvent` and making a full signature for that. If `ContractEvent` changes we change the signature and we have not made any separation real.
We just insisted on the logical separation, which is not there today at different levels.
Once we solve the whole picture it's even possible the client of the VM has a way to plug native functions and all of that disappears from the VM code and the first thing you do is to make that back to a `ContractEvent`.

Again though, I am fine either way, let's not just focus on what would be right, if that "right" is not really reachable with the timeline we have

Reply:
Yeah I think what you said is an accurate summarization of the current situation and I agree we don't have an event story for the Move VM yet. 

I'm also fine reverting this tuple to `ContractEvent`. I'll take another look during the final clean up of this PR and see what would be the cleanest option at that time.

Reply:
Let me move `ContractEvent` into move core types and get rid of the tuple.